jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -1528,15 +1540,47 @@ Method getMethodHelper(
 	 * Method N with return type S is more specific than M with return type R if:
 	 * S is the same as or a subtype of R.
 	 * Otherwise, the result method is chosen arbitrarily from specific methods.
+	 *
+	 * Note:
+	 * The more specific method is retrieved via the superclass rather than the direct interface
+	 * of this Class only if the method is also declared by another interface(retrieved via the superclass)
+	 * which is not the same or super-interface of the direct interface of this Class.
 	 */
-	bestCandidate = result;
+	bestCandidate = getMostSpecificMethod(forDeclaredMethod, methodList, result, name, strSig, parameterTypes);
+	
+	if (!forDeclaredMethod && (resultFromSuperClass != null)) {
+		candidateFromSuperClass = getMostSpecificMethod(false, methodList, resultFromSuperClass, name, strSig, parameterTypes);
```

Comment:
The callers of this helper method only specifies a non-null `methodList` when `forDeclaredMethod` is `true`. Hence `getMostSpecificMethod(false, methodList)` seems not necessary.  Further more, `getMostSpecificMethod` could add methods to `methodList` when `forDeclaredMethod` is `false` which is not correct.

Reply:
Discussed with @ChengJin01 , the second `getMostSpecificMethod` was within code block `!forDeclaredMethod` so `methodList` is `null` and won't be updated.
After all the new helper method `getMostSpecificMethod` is recipe for confusion, suggested to create a specific method name and with a clear purpose for super class.

Reply:
As discussed with @JasonFengJ9, to make the code clearer,  the original code there remains unchanged plus a helper method to deal with the most specific method from the super class.